File: homework3/Python/Program2.py
from enum import Enum


class Status(Enum):
    PENDING = 1
    APPROVED = 2
    REJECTED = 3
    
    
    # CANCELLED = 4  






def StatusComplete(status: Status) -> str:

    match status:
        case Status.PENDING:
            return "Status is PENDING - waiting for review"
        case Status.APPROVED:
            return "Status is APPROVED - action taken"
        case Status.REJECTED:
            return "Status is REJECTED - request denied"
        
        
        # case Status.CANCELLED:
        #     return "Status is CANCELLED - withdrawn"




def StatusIncomplete(status: Status) -> str:
    match status:
        case Status.PENDING:
            return "Status is PENDING - waiting for review"
        case Status.APPROVED:
            return "Status is APPROVED - action taken"
        
        # missing REJECTED case
        # case Status.REJECTED:
        #     return "Status is REJECTED - request denied"





# The main function goes through all the values of the Status enum. There are 2 parts, one where the function is complete and goes through all the cases and the other is missing 
# enum REJECTED case. Even if the REJECTED case is missing, the program will still run without any compilation errors because Python does not check for problems during compilation. 
# However, during execution/runtime, Python will raise a RuntimeError when it encounters the missing case because said case is not handled in the StatusIncomplete Function. 
def main():
    print("\n")
    print("PART 1: Exhaustive match")

    
    for status in Status:
        try:
            result = StatusComplete(status)
            print(f"{status.name}: {result}")
        except Exception as e:
            print(f"{status.name}: ERROR - {type(e).__name__}: {e}")
    
    print("\n")
    print("PART 2: Incomplete match)")
    
    for status in Status:
        try:
            result = StatusIncomplete(status)
            print(f"{status.name}: {result}")
        except Exception as e:
            print(f"{status.name}: RUNTIME ERROR - {type(e).__name__}: {e}")
    
    print("\n")

main()

File: homework3/Csharp/Program2.cs
using System;

namespace Homework3.Csharp
{
    // Task 2: Exhaustive Control Flow with enums
    internal static class Program2
    {
        // C# enum can only be value-types such as int, long, and byte. By default int is used.
        private enum Status
        {
            Pending = 1,
            Approved = 2,
            Rejected = 3,
            Cancelled = 4
        }

        private static string StatusWithSwitchStatementIncomplete(Status status)
        {
            // Intentionally missing Status.Rejected and Status.Cancelled.
            // C#'s traditional switch statement does NOT require exhaustive handling.
            // Any unhandled values must be caught using a default branch.

            switch (status)
            {
                case Status.Pending:
                    return "Status is PENDING - waiting for review";
                case Status.Approved:
                    return "Status is APPROVED - action taken";
                default:
                    return $"UNHANDLED STATUS: {status} ({(int)status})";
            }
        }

        private static string StatusWithSwitchExpressionComplete(Status status) => status switch
        {
            Status.Pending => "Status is PENDING",
            Status.Approved => "Status is APPROVED",
            Status.Rejected => "Status is REJECTED",
            Status.Cancelled => "Status is CANCELLED"
            // If a new Status value is added to the enum above and not handled here,
            // this switch expression will cause a compile-time error until updated.
        };

        private static void Main()
        {
            Console.WriteLine();
            Console.WriteLine("Exhaustive with complete Switch Expression:");

            foreach (Status status in Enum.GetValues<Status>())
            {
                string result = StatusWithSwitchExpressionComplete(status);
                
                Console.WriteLine($"\t{status}: {result}");
            }

            Console.WriteLine();
            Console.WriteLine("Exhaustive with incomplete, defaulted Switch Statement:");

            foreach (Status status in Enum.GetValues<Status>())
            {
                string result = StatusWithSwitchStatementIncomplete(status);
                
                Console.WriteLine($"\t{status}: {result}");
            }
        }
    }
}
